src/commands/trigger.cpp

```diff
@@ -0,0 +1,728 @@
+    if (predicate_ != nullptr) {
+      LOG_DEBUG("predicate_ is not nullptr");
+      LOG_DEBUG("predicate_ type = %s", ExpressionTypeToString(predicate_->GetExpressionType()).c_str());
+      LOG_DEBUG("predicate_ size = %lu", predicate_->GetChildrenSize());
+      if (executor_context_ != nullptr) {
+        LOG_DEBUG("before evalulate");
+        auto tuple_new = (const AbstractTuple *) tuple;
+        auto eval = predicate_->Evaluate(tuple_new, nullptr, executor_context_);
+        LOG_DEBUG("Evaluation result: %s", eval.GetInfo().c_str());
+        if (eval.IsTrue()) {
+          LOG_DEBUG("pass one trigger fire condition!!!");
+          LOG_DEBUG("trigger %s is fired!", triggers[i].GetTriggerName().c_str());
+        } else {
+          LOG_DEBUG("fail one trigger fire condition!!!");
+          continue;
+        }
+      }
+    } else {
+      LOG_DEBUG("predicate_ is nullptr");
+    }
+
+    // Construct trigger data
+    TriggerData trigger_data(trigger_type, &obj, nullptr, nullptr);
+
+    // apply all per-row-before-delete triggers on the table
+    obj.ExecCallTriggerFunc(trigger_data);
+  }
+  return true;
+}
+
+bool TriggerList::ExecARDeleteTriggers(storage::Tuple *tuple, executor::ExecutorContext *executor_context_) {
+  LOG_DEBUG("enter into ExecARDeleteTriggers");
+
+  //check valid type
+  if (!types_summary[EnumTriggerType::AFTER_DELETE_ROW]) {
+    return false;
+  }
+  unsigned i;
+  for (i = 0; i < triggers.size(); i++) {
+    Trigger &obj = triggers[i];
+    int16_t trigger_type = obj.GetTriggerType();
+    //check valid type
+    if (!TRIGGER_TYPE_MATCHES(trigger_type, TRIGGER_TYPE_ROW, TRIGGER_TYPE_AFTER, TRIGGER_TYPE_DELETE)) {
+      continue;
+    }
+
+    expression::AbstractExpression *predicate_ = obj.GetTriggerWhen();
+    if (predicate_ != nullptr) {
+      LOG_DEBUG("predicate_ is not nullptr");
+      LOG_DEBUG("predicate_ type = %s", ExpressionTypeToString(predicate_->GetExpressionType()).c_str());
+      LOG_DEBUG("predicate_ size = %lu", predicate_->GetChildrenSize());
+      if (executor_context_ != nullptr) {
+        LOG_DEBUG("before evalulate");
+        auto tuple_new = (const AbstractTuple *) tuple;
+        auto eval = predicate_->Evaluate(tuple_new, nullptr, executor_context_);
+        LOG_DEBUG("Evaluation result: %s", eval.GetInfo().c_str());
+        if (eval.IsTrue()) {
+          LOG_DEBUG("pass one trigger fire condition!!!");
+          LOG_DEBUG("trigger %s is fired!", triggers[i].GetTriggerName().c_str());
+        } else {
+          LOG_DEBUG("fail one trigger fire condition!!!");
+          continue;
+        }
+      }
+    } else {
+      LOG_DEBUG("predicate_ is nullptr");
+    }
+
+    // Construct trigger data
+    TriggerData trigger_data(trigger_type, &obj, nullptr, nullptr);
+
+    // apply all per-row-after-delete triggers on the table
+    obj.ExecCallTriggerFunc(trigger_data);
+  }
+  return true;
+}
+
+void TriggerList::ExecBSInsertTriggers() {
```

Comment:
I feel like we can combine a bunch of these Exec methods. We are repeating logic all over the place.